crates/miden-objects/src/block/nullifier_tree.rs

```diff
@@ -0,0 +1,297 @@
+/// not change. Note that inserting the nullifier multiple times with the same block number is
+/// valid.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct NullifierTree {
+    smt: Smt,
+}
+
+impl NullifierTree {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+
+    /// The depth of the nullifier tree.
+    pub const DEPTH: u8 = SMT_DEPTH;
+
+    /// The value of an unspent nullifier in the tree.
+    pub const UNSPENT_NULLIFIER: Word = EMPTY_WORD;
+
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Creates a new, empty nullifier tree.
+    pub fn new() -> Self {
+        Self { smt: Smt::new() }
+    }
+
+    /// Construct a new nullifier tree from the provided entries.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// - the provided entries contain multiple block numbers for the same nullifier.
+    pub fn with_entries(
+        entries: impl IntoIterator<Item = (Nullifier, BlockNumber)>,
+    ) -> Result<Self, NullifierTreeError> {
+        let leaves = entries.into_iter().map(|(nullifier, block_num)| {
+            (nullifier.inner(), Self::block_num_to_leaf_value(block_num))
+        });
+
+        let smt = Smt::with_entries(leaves)
+            .map_err(NullifierTreeError::DuplicateNullifierBlockNumbers)?;
+
+        Ok(Self { smt })
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns the root of the nullifier SMT.
+    pub fn root(&self) -> RpoDigest {
+        self.smt.root()
+    }
+
+    /// Returns the number of spent nullifiers in this tree.
+    pub fn num_nullifiers(&self) -> usize {
+        self.smt.num_entries()
+    }
+
+    /// Returns a [`NullifierWitness`] of the leaf associated with the `nullifier`.
+    ///
+    /// Conceptually, such a witness is a Merkle path to the leaf, as well as the leaf itself.
+    ///
+    /// This witness is a proof of the current block number of the given nullifier. If that block
+    /// number is zero, it proves that the nullifier is unspent.
+    pub fn open(&self, nullifier: &Nullifier) -> NullifierWitness {
+        NullifierWitness::new(self.smt.open(&nullifier.inner()))
+    }
+
+    /// Returns the block number for the given nullifier or `None` if the nullifier wasn't spent
+    /// yet.
+    pub fn get_block_num(&self, nullifier: &Nullifier) -> Option<BlockNumber> {
+        let value = self.smt.get_value(&nullifier.inner());
+        if value == Smt::EMPTY_VALUE {
+            return None;
+        }
+
+        Some(Self::leaf_value_to_block_num(value))
+    }
+
+    /// Computes mutations for the nullifier SMT.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// - a nullifier in the provided iterator was already spent.
```

Comment:
Should we clarify here that this only applies if two nullifiers were spent in different blocks?

Reply:
I'm not sure I fully understand. I think the docs accurately describe the current behavior. If any nullifier in the iterator was already spent, no matter at which block - even the same one, an error is returned, e.g.:

https://github.com/0xMiden/miden-base/blob/e315e930e02f3fbd0a6fe289460e5b1020e10550/crates/miden-objects/src/block/nullifier_tree.rs#L283-L305

Is that not the behavior we want? E.g. do we want to be able to successfully mark a nullifier as spent as long as it points to the same block as before?

Reply:
hmmm - i think I might have misunderstood. n/m!